Mathlib/GroupTheory/RegularWreathProduct.lean

```diff
@@ -0,0 +1,293 @@
+  mul_one a := RegularWreathProduct.ext (by simp; rfl) (mul_one _)
+  inv_mul_cancel a :=
+    RegularWreathProduct.ext (by simp; exact mul_eq_one_iff_eq_inv.mpr rfl) (by simp)
+
+instance : Inhabited (RegularWreathProduct D Q) := ⟨1⟩
+
+/-- The canonical projection map `D ≀ᵣ Q →* Q`, as a group hom. -/
+def rightHom : D ≀ᵣ Q →* Q where
+  toFun := RegularWreathProduct.right
+  map_one' := rfl
+  map_mul' _ _ := rfl
+
+/-- The canonical map `Q →* D ≀ᵣ Q` sending `q` to `⟨1, q⟩` -/
+def inl : Q →* D ≀ᵣ Q where
+  toFun q := ⟨1, q⟩
+  map_one' := rfl
+  map_mul' _ _ := by ext <;> simp
+
+@[simp]
+theorem left_inl (q : Q) : (inl q : D ≀ᵣ Q).left = 1 := rfl
+
+@[simp]
+theorem right_inl (q : Q) : (inl q : D ≀ᵣ Q).right = q := rfl
+
+@[simp]
+theorem rightHom_eq_right : (rightHom : D ≀ᵣ Q → Q) = right := rfl
+
+@[simp]
+theorem rightHom_comp_inl_eq_id : (rightHom : D ≀ᵣ Q →* Q).comp inl = MonoidHom.id _ := by ext; simp
+
+@[simp]
+theorem fun_id (q : Q) : rightHom (inl q : D ≀ᵣ Q) = q := by simp
+
+/-- The equivalence map for the representation as a product. -/
+def equivProd D Q : D ≀ᵣ Q ≃ (Q → D) × Q where
+  toFun := fun ⟨d, q⟩ => ⟨d, q⟩
+  invFun := fun ⟨d, q⟩ => ⟨d, q⟩
+  left_inv := fun _ => rfl
+  right_inv := fun _ => rfl
+
+omit [Group D] [Group Q] in
+lemma equivProdInj : Function.Injective (equivProd D Q).toFun := by
+  intro a b; simp
+
+instance [Finite D] [Finite Q] : Finite (D ≀ᵣ Q) := by
+  apply (Equiv.finite_iff (equivProd D Q)).mpr
+  apply Finite.instProd
+
+omit [Group D] [Group Q] in
+theorem card [Finite Q] :
+   Nat.card (D ≀ᵣ Q) = (Nat.card D^Nat.card Q) * Nat.card Q := by
+  rw [Nat.card_congr (equivProd D Q)]
+  rw [Nat.card_prod (Q → D) Q]
+  rw [Nat.card_fun]
+
+/-- Define an isomorphism from `D₁ ≀ᵣ Q₁` to `D₂ ≀ᵣ Q₂`
+given isomorphisms `D₁ ≀ᵣ Q₁` and `Q₁ ≃* Q₂`. -/
+def congr {D₁ Q₁ D₂ Q₂ : Type*} [Group D₁] [Group Q₁] [Group D₂] [Group Q₂]
+    (f : D₁ ≃* D₂) (g : Q₁ ≃* Q₂) :
+    D₁ ≀ᵣ Q₁ ≃* D₂ ≀ᵣ Q₂ where
+      toFun x := ⟨f ∘ (x.left ∘ g.symm), g x.right⟩
+      invFun x := ⟨(f.symm ∘ x.left) ∘ g, g.symm x.right⟩
+      left_inv x := by ext <;> simp
+      right_inv x := by ext <;> simp
+      map_mul' x y := by ext <;> simp
+
+section perm
+
+variable (D : Type*) [Group D]
+variable (Q : Type*) [Group Q] [Nonempty Q]
+variable (Λ : Type*) [Nonempty Λ] [MulAction D Λ]
+
+instance instSMulRWP : SMul (D ≀ᵣ Q) (Λ × Q) where
+  smul w := fun p => ⟨(w.left (w.right * p.2)) • p.1, w.right * p.2⟩
+
+omit [Nonempty Λ] [Nonempty Q] in
+@[simp]
+lemma rsmul {w : D ≀ᵣ Q} {p : Λ × Q} :
+    w • p = ⟨(w.left (w.right * p.2)) • p.1, w.right * p.2⟩ := rfl
+
+instance instMulActionRWP : MulAction (D ≀ᵣ Q) (Λ × Q) where
+  one_smul := by simp
+  mul_smul := by simp; intro x y a b; constructor
+                 · rw [smul_smul]; group
+                 · exact mul_assoc x.right y.right b
+
+
+variable [FaithfulSMul D Λ]
+instance instFaithfulSMulRWP : FaithfulSMul (D ≀ᵣ Q) (Λ × Q) where
+  eq_of_smul_eq_smul := by
+    simp
```

Comment:
We discourage `simp` in the middle of a proof since if the simp-set is changed, then the proof can break unexpectedly. Instead, squeeze this simp:
```suggestion
    simp only [rsmul, Prod.mk.injEq, mul_left_inj, Prod.forall]
```